presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java

```diff
@@ -0,0 +1,176 @@
+
+    private final long fileSize;
+    private final byte[] tailBuffer;
+    private final int length;
+
+    private FSDataInputStreamTail(long fileSize, byte[] tailBuffer, int length)
+    {
+        this.tailBuffer = requireNonNull(tailBuffer, "tailBuffer is null");
+        this.fileSize = fileSize;
+        this.length = length;
+        checkArgument(fileSize >= 0, "fileSize must be >= 0, found: %s", fileSize);
+        checkArgument(length <= fileSize, "length must be <= fileSize, found: %s > %s", length, fileSize);
+        checkPositionIndexes(0, length, tailBuffer.length);
+    }
+
+    public static FSDataInputStreamTail readTail(Path path, long paddedFileSize, FSDataInputStream inputStream, int minimumLength)
+            throws IOException
+    {
+        checkArgument(minimumLength >= 0 && minimumLength <= MAXIMUM_MIN_READ_LENGTH, "minimumLength too large, found: %s", minimumLength);
+        long readSize = min(paddedFileSize, (minimumLength + ADDITIONAL_READ_PADDING_BYTES));
+        long position = paddedFileSize - readSize;
+        // Actual read will be 1 byte larger to ensure we encounter an EOF where expected
+        byte[] buffer = new byte[toIntExact(readSize + 1)];
+        int bytesRead = inputStream.read(position, buffer, 0, buffer.length);
+        if (bytesRead > readSize) {
+            throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Incorrect fileSize %s for file %s, end of stream not reached at offset %s", paddedFileSize, path, (position + bytesRead)));
+        }
+        return new FSDataInputStreamTail(position + bytesRead, buffer, bytesRead);
+    }
+
+    public long getFileSize()
+    {
+        return fileSize;
+    }
+
+    public Slice getTailSlice()
+    {
+        return Slices.wrappedBuffer(tailBuffer, 0, length);
+    }
+
+    public FSDataInputStream replaceStreamWithContentsIfComplete(FSDataInputStream originalStream)
+            throws IOException
+    {
+        if (fileSize == length) {
+            originalStream.close();
+            return new FSDataInputStream(new ByteArraySeekablePositionedReadableInputStream(tailBuffer, 0, length));
+        }
+        return originalStream;
+    }
+
+    @VisibleForTesting
+    static final class ByteArraySeekablePositionedReadableInputStream
+            extends ByteArrayInputStream
+            implements Seekable, PositionedReadable
+    {
+        private ByteArraySeekablePositionedReadableInputStream(byte[] array, int offset, int length)
+        {
+            super(array, offset, length);
+        }
+
+        @Override
+        public void seek(long pos)
+                throws IOException
+        {
+            if (pos < 0) {
+                throw new IOException(FSExceptionMessages.NEGATIVE_SEEK);
+            }
+            if (pos > count) {
+                throw new IOException(FSExceptionMessages.CANNOT_SEEK_PAST_EOF);
+            }
+            this.reset();
```

Comment:
This won't work if someone has called `mark()` on the stream. But `pos` is protected so you can modify it directly.